### PR TITLE
SONARJAVA-2309 Symbolic values for resources should be added transiti…

### PR DESCRIPTION
**Pull request by saberduck**
…vely

### CODE REVIEW

File: java-frontend/src/main/java/org/sonar/java/se/ProgramState.java [hunk 1]
```diff
@@ -259,6 +260,22 @@ public String toString() {
     return "{" + values.toString() + "}  {" + constraints.toString() + "}" + " { " + stack.toString() + " }";
   }
 
+  public ProgramState addConstraintTransitively(SymbolicValue symbolicValue, Constraint constraint) {
+    final ProgramState[] ps = {addConstraint(symbolicValue, constraint)};
+    knownRelations()
+      .filter(rsv -> rsv.isEquality() && (rsv.getLeftOp() == symbolicValue || rsv.getRightOp() == symbolicValue))
+      .map(rsv -> rsv.getLeftOp() == symbolicValue ? rsv.getRightOp() : rsv.getLeftOp())
+      .forEach(sv -> ps[0] = ps[0].addConstraint(sv, constraint));
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/se/ProgramState.java]: I am pretty sure what we are looking for here is a collector acting as reduction : https://docs.oracle.com/javase/tutorial/collections/streams/reduction.html

While this final array works it is not really nice.

File: java-frontend/src/main/java/org/sonar/java/se/ProgramState.java [hunk 2]
```diff
@@ -259,6 +260,22 @@ public String toString() {
     return "{" + values.toString() + "}  {" + constraints.toString() + "}" + " { " + stack.toString() + " }";
   }
 
+  public ProgramState addConstraintTransitively(SymbolicValue symbolicValue, Constraint constraint) {
+    final ProgramState[] ps = {addConstraint(symbolicValue, constraint)};
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/se/ProgramState.java]: For sake of clarity i would separate array initializer and method call.

File: java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/RelationalSymbolicValue.java [hunk 3]
```diff
@@ -430,7 +430,7 @@ private boolean isCommutative() {
     }
   }
 
-  private boolean isEquality() {
+  public boolean isEquality() {
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/RelationalSymbolicValue.java]: I would add a dedicated test if this is made public.

File: java-frontend/src/main/java/org/sonar/java/se/ProgramState.java (hunk 1 above)
Comment (saberduck) [inline java-frontend/src/main/java/org/sonar/java/se/ProgramState.java]: Unfortunately, stream API doesn't provide fold-left operation, and reduce method would require combiner to allow merging of two intermediate ProgramStates. But I understand that abusing arrays for this is bit of a hack, so I rewrote it by having values collected.

**Pull request merged**